File: cc/aead/kms_aead_key_manager_test.cc
// Copyright 2019 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

#include "tink/aead/kms_aead_key_manager.h"

#include "gtest/gtest.h"
#include "absl/status/status.h"
#include "absl/strings/match.h"
#include "absl/strings/str_cat.h"
#include "tink/aead.h"
#include "tink/kms_client.h"
#include "tink/kms_clients.h"
#include "tink/subtle/aead_test_util.h"
#include "tink/util/status.h"
#include "tink/util/statusor.h"
#include "tink/util/test_matchers.h"
#include "tink/util/test_util.h"
#include "proto/kms_aead.pb.h"

namespace crypto {
namespace tink {

using ::crypto::tink::test::DummyAead;
using ::crypto::tink::test::DummyKmsClient;
using ::crypto::tink::test::IsOk;
using ::crypto::tink::test::StatusIs;
using ::google::crypto::tink::KmsAeadKey;
using ::google::crypto::tink::KmsAeadKeyFormat;
using ::testing::Eq;
using ::testing::Not;

namespace {

TEST(KmsAeadKeyManagerTest, Basics) {
  EXPECT_THAT(KmsAeadKeyManager().get_version(), Eq(0));
  EXPECT_THAT(KmsAeadKeyManager().get_key_type(),
              Eq("type.googleapis.com/google.crypto.tink.KmsAeadKey"));
  EXPECT_THAT(KmsAeadKeyManager().key_material_type(),
              Eq(google::crypto::tink::KeyData::REMOTE));
}

TEST(KmsAeadKeyManagerTest, ValidateEmptyKey) {
  EXPECT_THAT(KmsAeadKeyManager().ValidateKey(KmsAeadKey()),
              StatusIs(absl::StatusCode::kInvalidArgument));
}

TEST(KmsAeadKeyManagerTest, ValidateValidKey) {
  KmsAeadKey key;
  key.set_version(0);
  key.mutable_params()->set_key_uri("Some uri");
  EXPECT_THAT(KmsAeadKeyManager().ValidateKey(key), IsOk());
}

TEST(KmsAeadKeyManagerTest, ValidateWrongVersion) {
  KmsAeadKey key;
  key.set_version(1);
  key.mutable_params()->set_key_uri("Some uri");
  EXPECT_THAT(KmsAeadKeyManager().ValidateKey(key), Not(IsOk()));
}

TEST(KmsAeadKeyManagerTest, ValidateNoUri) {
  KmsAeadKey key;
  key.set_version(0);
  EXPECT_THAT(KmsAeadKeyManager().ValidateKey(key), Not(IsOk()));
}

TEST(KmsAeadKeyManagerTest, ValidateKeyFormatEmptyKey) {
  EXPECT_THAT(KmsAeadKeyManager().ValidateKeyFormat(KmsAeadKeyFormat()),
              StatusIs(absl::StatusCode::kInvalidArgument));
}

TEST(KmsAeadKeyManagerTest, ValidateKeyFormatValidKey) {
  KmsAeadKeyFormat key_format;
  key_format.set_key_uri("Some uri");
  EXPECT_THAT(KmsAeadKeyManager().ValidateKeyFormat(key_format), IsOk());
}

TEST(KmsAeadKeyManagerTest, ValidateKeyFormatNoUri) {
  KmsAeadKeyFormat key_format;
  EXPECT_THAT(KmsAeadKeyManager().ValidateKeyFormat(key_format), Not(IsOk()));
}

TEST(KmsAeadKeyManagerTest, CreateKey) {
  KmsAeadKeyFormat key_format;
  key_format.set_key_uri("Some uri");
  auto key_or = KmsAeadKeyManager().CreateKey(key_format);
  ASSERT_THAT(key_or.status(), IsOk());
  EXPECT_THAT(key_or.ValueOrDie().params().key_uri(), Eq(key_format.key_uri()));
}

class KmsAeadKeyManagerCreateTest : public ::testing::Test {
 public:
  // The KmsClients class has a global variable which keeps the registered
  // clients. To reflect that in the test, we set them up in the SetUpTestSuite
  // function.
  static void SetUpTestSuite() {
    if (!KmsClients::Add(
             absl::make_unique<DummyKmsClient>("prefix1", "prefix1:some_key1"))
             .ok())
      abort();
    if (!KmsClients::Add(absl::make_unique<DummyKmsClient>("prefix2", "")).ok())
      abort();
  }
};

TEST_F(KmsAeadKeyManagerCreateTest, CreateAead) {
  KmsAeadKey key;
  key.set_version(0);
  key.mutable_params()->set_key_uri("prefix1:some_key1");

  auto kms_aead = KmsAeadKeyManager().GetPrimitive<Aead>(key);
  ASSERT_THAT(kms_aead.status(), IsOk());

  DummyAead direct_aead("prefix1:some_key1");

  EXPECT_THAT(EncryptThenDecrypt(*kms_aead.ValueOrDie(), direct_aead,
                                 "plaintext", "aad"),
              IsOk());
}

TEST_F(KmsAeadKeyManagerCreateTest, CreateAeadWrongKeyName) {
  KmsAeadKey key;
  key.set_version(0);
  key.mutable_params()->set_key_uri("prefix1:some_other_key");

  auto kms_aead = KmsAeadKeyManager().GetPrimitive<Aead>(key);
  ASSERT_THAT(kms_aead.status(), Not(IsOk()));
}

TEST_F(KmsAeadKeyManagerCreateTest, CreateAeadWrongPrefix) {
  KmsAeadKey key;
  key.set_version(0);
  key.mutable_params()->set_key_uri("non-existing-prefix:some_key1");

  auto kms_aead = KmsAeadKeyManager().GetPrimitive<Aead>(key);
  ASSERT_THAT(kms_aead.status(), Not(IsOk()));
}

TEST_F(KmsAeadKeyManagerCreateTest, CreateAeadUnboundKey) {
  KmsAeadKey key;
  key.set_version(0);
  key.mutable_params()->set_key_uri("prefix2:some_key2");

  auto kms_aead = KmsAeadKeyManager().GetPrimitive<Aead>(key);
  ASSERT_THAT(kms_aead.status(), IsOk());

  DummyAead direct_aead("prefix2:some_key2");

  EXPECT_THAT(EncryptThenDecrypt(*kms_aead.ValueOrDie(), direct_aead,
                                 "plaintext", "aad"),
              IsOk());
}

}  // namespace
}  // namespace tink
}  // namespace crypto

File: python/tink/aead/_kms_aead_key_manager_test.py
# Copyright 2023 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Tests for kms_aead_key_manager."""

from absl.testing import absltest
from tink.proto import kms_aead_pb2
from tink.proto import kms_envelope_pb2
from tink.proto import tink_pb2
import tink
from tink import aead
from tink.aead import _kms_aead_key_manager


def setUpModule():
  aead.register()


class FakeClient(tink.KmsClient):

  def __init__(self, key_uri):
    self.key_uri = key_uri

  def does_support(self, key_uri: str) -> bool:
    return key_uri == self.key_uri

  def get_aead(self, key_uri: str) -> aead.Aead:
    raise NotImplementedError('not implemented')


KMS_AEAD_KEY_TYPE_URL = 'type.googleapis.com/google.crypto.tink.KmsAeadKey'
KMS_ENVELOPE_AEAD_KEY_TYPE_URL = (
    'type.googleapis.com/google.crypto.tink.KmsEnvelopeAeadKey'
)


class KmsAeadKeyManagerTest(absltest.TestCase):

  def test_kms_aead_new_key_data_success(self):
    client = FakeClient('key_uri')
    tink.register_kms_client(client)
    template = aead.aead_key_templates.create_kms_aead_key_template(
        key_uri='key_uri'
    )
    key_data = _kms_aead_key_manager.KmsAeadKeyManager().new_key_data(template)

    self.assertEqual(key_data.type_url, KMS_AEAD_KEY_TYPE_URL)
    self.assertEqual(key_data.key_material_type, tink_pb2.KeyData.REMOTE)
    key = kms_aead_pb2.KmsAeadKey.FromString(key_data.value)
    self.assertEqual(key.version, 0)
    self.assertEqual(key.params.key_uri, 'key_uri')

  def test_kms_aead_new_key_data_rejects_unknown_template(self):
    with self.assertRaises(tink.TinkError):
      _kms_aead_key_manager.KmsAeadKeyManager().new_key_data(
          aead.aead_key_templates.XCHACHA20_POLY1305_RAW
      )

  def test_kms_aead_primitive_rejects_unknown_key_data(self):
    template = aead.aead_key_templates.create_kms_envelope_aead_key_template(
        kek_uri='kek_uri',
        dek_template=aead.aead_key_templates.XCHACHA20_POLY1305_RAW,
    )
    envelope_aead_key_data = (
        _kms_aead_key_manager.KmsEnvelopeAeadKeyManager().new_key_data(template)
    )

    with self.assertRaises(tink.TinkError):
      _kms_aead_key_manager.KmsAeadKeyManager().primitive(
          envelope_aead_key_data
      )

  def test_kms_envelope_aead_new_key_data_success(self):
    template = aead.aead_key_templates.create_kms_envelope_aead_key_template(
        kek_uri='kek_uri',
        dek_template=aead.aead_key_templates.XCHACHA20_POLY1305_RAW,
    )
    key_data = _kms_aead_key_manager.KmsEnvelopeAeadKeyManager().new_key_data(
        template
    )

    self.assertEqual(key_data.type_url, KMS_ENVELOPE_AEAD_KEY_TYPE_URL)
    self.assertEqual(key_data.key_material_type, tink_pb2.KeyData.REMOTE)
    key = kms_envelope_pb2.KmsEnvelopeAeadKey.FromString(key_data.value)
    self.assertEqual(key.version, 0)
    self.assertEqual(key.params.kek_uri, 'kek_uri')
    self.assertEqual(
        key.params.dek_template, aead.aead_key_templates.XCHACHA20_POLY1305_RAW
    )

  def test_kms_envelope_aead_rejects_unknown_template(self):
    with self.assertRaises(tink.TinkError):
      _kms_aead_key_manager.KmsEnvelopeAeadKeyManager().new_key_data(
          aead.aead_key_templates.XCHACHA20_POLY1305_RAW
      )

  def test_kms_envelope_aead_primitive_rejects_unknown_key_data(self):
    template = aead.aead_key_templates.create_kms_aead_key_template(
        key_uri='key_uri'
    )
    kms_aead_key_data = _kms_aead_key_manager.KmsAeadKeyManager().new_key_data(
        template
    )

    with self.assertRaises(tink.TinkError):
      _kms_aead_key_manager.KmsEnvelopeAeadKeyManager().primitive(
          kms_aead_key_data
      )


if __name__ == '__main__':
  absltest.main()
